# 1. Load the datasets

In [1]:
import numpy as np
import pandas as pd

In [15]:
data1_test = pd.read_csv('./tweet_dataset_1/test.csv')
data1_train = pd.read_csv('./tweet_dataset_1/training.csv')
data1_validate = pd.read_csv('./tweet_dataset_1/validation.csv')

data1 = pd.concat([data1_test, data1_train, data1_validate], ignore_index= True)

data2 = pd.read_csv('./tweet_dataset_2/tweet_emotions.csv')

print("Data 1 size : {}".format(data1.shape))
print("Data 2 size : {}".format(data2.shape))


Data 1 size : (20000, 2)
Data 2 size : (40000, 3)


In [16]:
df1 = pd.DataFrame(data1)
df2 = pd.DataFrame(data2)

# 2. Checkout dataset 1

In [17]:
df1.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 20000 entries, 0 to 19999
Data columns (total 2 columns):
 #   Column  Non-Null Count  Dtype 
---  ------  --------------  ----- 
 0   text    20000 non-null  object
 1   label   20000 non-null  int64 
dtypes: int64(1), object(1)
memory usage: 312.6+ KB


In [18]:
df1['label'].unique()

array([0, 1, 4, 3, 2, 5], dtype=int64)

> <b>WE KNOW THAT</b> : possible values including:<br>
- sadness(0) <br>
- joy(1) <br>
- love(2) <br>
- anger(3) <br>
- fear (4) <br>
- (5) unknown????? (IDENTIFY) <br>

In [19]:
df1_label_5 = df1[df1['label']== 5].reset_index()
df1_label_5.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 719 entries, 0 to 718
Data columns (total 3 columns):
 #   Column  Non-Null Count  Dtype 
---  ------  --------------  ----- 
 0   index   719 non-null    int64 
 1   text    719 non-null    object
 2   label   719 non-null    int64 
dtypes: int64(2), object(1)
memory usage: 17.0+ KB


In [20]:
df1_label_5['text'][490]

'i love seeing what books resonate with my girls i love seeing their faces grow serious when characters face complications trials and obstacles and i love the discussions that come out of reading time as we talk about main ideas how the books made us feel and what may have surprised us'

<p> cannot identify label (5) so dropping it.<br> Only has 719 entries so wont affect much</p>

In [21]:
df1 = df1[df1['label'] != 5]
df1.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 19281 entries, 0 to 19999
Data columns (total 2 columns):
 #   Column  Non-Null Count  Dtype 
---  ------  --------------  ----- 
 0   text    19281 non-null  object
 1   label   19281 non-null  int64 
dtypes: int64(1), object(1)
memory usage: 451.9+ KB


# 3. Checkout dataset 2

In [22]:
df2.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 40000 entries, 0 to 39999
Data columns (total 3 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   tweet_id   40000 non-null  int64 
 1   sentiment  40000 non-null  object
 2   content    40000 non-null  object
dtypes: int64(1), object(2)
memory usage: 937.6+ KB


In [23]:
print(df2['sentiment'].unique())
print(df2['sentiment'].value_counts().unique())

['empty' 'sadness' 'enthusiasm' 'neutral' 'worry' 'surprise' 'love' 'fun'
 'hate' 'happiness' 'boredom' 'relief' 'anger']
[8638 8459 5209 5165 3842 2187 1776 1526 1323  827  759  179  110]


In [24]:
df2_label_empty = df2['content'][df2['sentiment']== 'empty'].reset_index()
df2_label_surprise = df2['content'][df2['sentiment']== 'surprise'].reset_index()
df2_label_relief = df2['content'][df2['sentiment']== 'relief'].reset_index()

In [25]:
df2_label_empty['content'][20]

"@allyycase I'm trying to sleep!!! It's not working"

In [26]:
df2_label_surprise['content'][230]

"NO! Our TV's busted  The screen turned white for about a nanosecond while it made a quick &quot;popping&quot; sound."

In [27]:
df2_label_relief['content'][776]

'Drinking tea and eating choc chip cookies. Going to head to sleep soon! Night all'

dataset 1:       --     dataset 2:
- sadness(0)     --     sadness (boredom, surprise)<br> 
- joy(1)         --     enthusiasm, happiness, fun (neutral)<br>
- love(2)        --     love<br>
- anger(3)       --     hate, anger (empty, relief)<br>
- fear (4)       --     worry, <br><br>

DATASET 2 leftout : 

# 4. Combine datasets
<p> convert dataset 2 content into dataset 1 numbers as per emotion mentioned above</p>

In [28]:
# before changes are made
df2.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 40000 entries, 0 to 39999
Data columns (total 3 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   tweet_id   40000 non-null  int64 
 1   sentiment  40000 non-null  object
 2   content    40000 non-null  object
dtypes: int64(1), object(2)
memory usage: 937.6+ KB


In [29]:
# drop tweet_id column
df2.drop(['tweet_id'], axis = 1, inplace = True)

# replace content column with numeric values mapped to dataset 1
df2['sentiment'].replace(['sadness', 'boredom', 'surprise'], 0, inplace = True)
df2['sentiment'].replace(['enthusiasm', 'happiness', 'fun', 'neutral'], 1, inplace = True)
df2['sentiment'].replace(['love'], 2, inplace = True)
df2['sentiment'].replace(['hate', 'anger', 'empty', 'relief'], 3, inplace = True)
df2['sentiment'].replace(['worry'], 4, inplace = True)

In [30]:
# after changes are made
df2.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 40000 entries, 0 to 39999
Data columns (total 2 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   sentiment  40000 non-null  int64 
 1   content    40000 non-null  object
dtypes: int64(1), object(1)
memory usage: 625.1+ KB


In [31]:
df2['sentiment'].unique()

array([3, 0, 1, 4, 2], dtype=int64)

In [32]:
df2.rename(columns = {'sentiment' : 'label', 'content' : 'text'}, inplace = True)
df2.head()

label                                               text
0      3  @tiffanylue i know  i was listenin to bad habi...
1      0  Layin n bed with a headache  ughhhh...waitin o...
2      0                Funeral ceremony...gloomy friday...
3      1               wants to hang out with friends SOON!
4      1  @dannycastillo We want to trade with someone w...

<b>Compare both dataframes</b>

In [33]:
df1.head()

text  label
0  im feeling rather rotten so im not very ambiti...      0
1          im updating my blog because i feel shitty      0
2  i never make her separate from me because i do...      0
3  i left with my bouquet of red and yellow tulip...      1
4    i was feeling a little vain when i did this one      0

In [34]:
df2.head()

label                                               text
0      3  @tiffanylue i know  i was listenin to bad habi...
1      0  Layin n bed with a headache  ughhhh...waitin o...
2      0                Funeral ceremony...gloomy friday...
3      1               wants to hang out with friends SOON!
4      1  @dannycastillo We want to trade with someone w...

In [35]:
df = pd.concat([df1, df2])
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 59281 entries, 0 to 39999
Data columns (total 2 columns):
 #   Column  Non-Null Count  Dtype 
---  ------  --------------  ----- 
 0   text    59281 non-null  object
 1   label   59281 non-null  int64 
dtypes: int64(1), object(1)
memory usage: 1.4+ MB


In [36]:
#!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
# USE IF U WANT TO SAVE THE COMBINED DATASET
#!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
df.to_csv("./DL_Dataset.csv", )

# 5. Analysis for Cleaning the dataset 
1. Need to remove unnecessary words from 'text' column

In [2]:
# if using new notebook for a saved dataset run this cell
df = pd.read_csv("./DL_Dataset.csv") # change name as per whatever you have saved it as
df.head()

Unnamed: 0                                               text  label
0           0  im feeling rather rotten so im not very ambiti...      0
1           1          im updating my blog because i feel shitty      0
2           2  i never make her separate from me because i do...      0
3           3  i left with my bouquet of red and yellow tulip...      1
4           4    i was feeling a little vain when i did this one      0

In [3]:
# drop Unnamed: 0 column 
df.drop(['Unnamed: 0'], axis = 1, inplace = True)
df.head()

text  label
0  im feeling rather rotten so im not very ambiti...      0
1          im updating my blog because i feel shitty      0
2  i never make her separate from me because i do...      0
3  i left with my bouquet of red and yellow tulip...      1
4    i was feeling a little vain when i did this one      0

In [4]:
len(df)

59281

In [40]:
first_word = list()
for i in range(len(df)):
    word = df['text'][i].split()[0]
    first_word.append(word)

first_word = pd.Series(first_word)
unique_first_word = first_word.unique()

In [41]:
# view result
print(len(unique_first_word))
for w in unique_first_word:
    print(w)

20984
im
i
ive
when
made
is
told
getting
whenever
always
ill
at
this
id
on
finding
fear
the
my
as
waiting
a
imdoing
while
we
sometime
there
in
stranded
after
earth
for
during
arriving
staying
watched
death
having
tutorial
heated
once
went
first
going
every
watching
boy
no
occured
being
iv
realizing
hearing
one
before
insulted
listening
day
four
out
lost
discovering
talking
someone
climbing
winning
that
deception
diagnosis
@tiffanylue
Layin
Funeral
wants
@dannycastillo
Re-pinging
I
Hmmm.
@charviray
@kelcouch
cant
Choked
Ugh!
@BrodyJenner
Got
The
@annarosekerr
So
@PerezHilton
How
@raaaaaaek
Wondering
No
so
On
@IsaacMascote
Damm
Fudge....
It
@cynthia_123
feels
SoCal!
Screw
@ether_radio
has
@GABBYiSACTiVE
mmm
@DavidArchie
just
Why
@creyes
Bed!!!!!...
@onscrn
I'm
@jakeboyd,
wnna
Chocolate
why
@djmicdamn
@Mennard
@benballer
sucks
bed...sorta.
@ramtops
@lostluna
@maternitytees
@itsgabbith
diesel
@Pokinatcha
Ok
At
@DavidCookLove
@vincew
...and
Brothers
says
Bitten
@Dancing_Monk
@havingmysay
@s

<b> too long to check.</b>
<p> What we need is starting words that are not alphabets</p>

In [49]:
print("ASCII for 'a': {} \nASCII for 'z': {} \nASCII for 'A': {} \nASCII for 'Z': {}". format(ord('a'), ord('z'), ord('A'), ord('Z')))
print("\n\nASCII for 0: {} \nASCII for 9: {}". format(ord('0'), ord('9')))

ASCII for 'a': 97 
ASCII for 'z': 122 
ASCII for 'A': 65 
ASCII for 'Z': 90


ASCII for 0: 48 
ASCII for 9: 57


In [43]:
non_alphabet_words = list()
for w in unique_first_word:
    ascii_val = ord(w[0])
    if (ascii_val < 65) or (ascii_val > 122) or (ascii_val > 90 and ascii_val < 97):
        non_alphabet_words.append(w)

In [44]:
# view result
print(len(non_alphabet_words))
for w in non_alphabet_words:
    print(w)

15461
@tiffanylue
@dannycastillo
@charviray
@kelcouch
@BrodyJenner
@annarosekerr
@PerezHilton
@raaaaaaek
@IsaacMascote
@cynthia_123
@ether_radio
@GABBYiSACTiVE
@DavidArchie
@creyes
@onscrn
@jakeboyd,
@djmicdamn
@Mennard
@benballer
@ramtops
@lostluna
@maternitytees
@itsgabbith
@Pokinatcha
@DavidCookLove
@vincew
...and
@Dancing_Monk
@havingmysay
@shondarhimes
@RachelLock22
@gcrush
@artfuldodga
@emmarler
..
@lepetitagneau
@willxxmobb
@RobertF3
@sweeetnspicy
@
@mrgenius23
@soviet_star
@cayogial
@mileycyrus
@justamedicine
@melbournegirl
@thecreativeone
@neesabear
@IamYeTe
@freepbx
@poinktoinkdoink
@britblackbird
@megturney
$#@!
*sigh*
@jwillock
@TheHarvardian
@LaFloozita
@softtouchme
@omfgiselle
@NisforNeemah
@Emilyyy16
2
@IdleThumbs
@nzdeany
@BarbSchaefer
@heresmyhello92
@juneyee
@jackgraycnn
@jaychuck
@relly1
@taxidermi
@JessiJaeJoplin
@jertronic
@TheLastDoctor
@Bern_morley
@xdjio
@melluffsyew
(@wendyisastar)
@AlexanderGWhite
@yuki_hime
@DymeDiva23
@rlcpbiatch
@SparkDawgMusic
@aiderushton

@Joseph_KHC
@DSTROYR
@JKeith1982
@irkengir
@Natalija
@kellanator
@wolfofmibu
@yelainez
@dopeboySHAKE
@BrentLauren
@pourmecoffee
@MooNooH
@J2thaESSICA
@Burgandy1
@SmellTheCheese
@ihatecrayons
@metafisix
@quikness
@kyza
@ghozali
@yellowbag
@Fasn8nUrMind
@LouGagliardi
@whitespider1066
@dj_cm
???��
@thomasfiss
@Nanava_3
@AdamKruth
@melisamelisa
@kim_flores
@jessereed
@forevertanoya
@tifftiff88
#DepecheMode
@Radio702
@avrilchan
@mcvie
@kevindavis1914
@Meikah
@kimoraklein
@Cause4Conceit
@philcampbell
@iancpalmer
@ErikVeland
@risamarie:
@busays12
2moro
@arania
@godfatherobie
@stevecruzblog
@ayawan
@greentm
@BARACKOKITTY
@Adamastorlx
@sashaostrikoff
@galaxydazzle
@_C_A_S_
@SharktailYPP
@deezeen
@lumivalkoinen
@SummerXoxo
@msalonen
@powertwitter
@charlietotem
@marjorielu
@BeckyBuckwild
@trinsgarcia
@ZDAEfron
@ahhrjay
@Jazzy813
@erikamodina
@digicho
@dancinqueenie
@EnglishRoses
@AroundHarlem
@shortstackband
@lisisilveira
@Emsy
@markbuchholz
@onmammaz
@brandillio
@pkBLGdonahue
@Maccah
???????

In [45]:
"""
# check why it didnt work later

words_to_keep = list
for w in non_alphabet_words:
    if w[0].startswith('@') == True:
        non_alphabet_words.remove(w)
    else:
        words_to_keep.append(w)
"""

words_to_keep = list()
for w in non_alphabet_words:
    if w[0].startswith('@') != True:
        words_to_keep.append(w)

In [12]:
print(len(words_to_keep))
print(len(non_alphabet_words))

213
7837


In [46]:
# view result
print(len(words_to_keep))
for w in words_to_keep:
    print(w)

416
...and
..
$#@!
*sigh*
2
(@wendyisastar)
#3wordsaftersex
25-man
0
#liesgirlstell
#beer
#jonaswebcast
4
2nd
&quot;my
...
2orw!!!!!
..i
3am
#thingsmummysaid
&quot;locked
???-??
???��
#DepecheMode
2moro
???????
1
:|
20
4am
5500
3)
3/4
365days
10
...I
?
3
#3breakupwords
#freakyfriday
&quot;
;
#Ferber
3rd
..the
3-4
[BBC]
#liesboystell
4th
29
33
#freeagentapp
1am,
#bigfanfriday
[Wrong!]
?�
&quot;sometime
:O
:
*sits
&quot;I'm
'm
#flylady
&quot;No
.@daveharte
(@zachflauaus)
25
&quot;yeah
'Wake
#asylm
14
'ss
1:13
#ff
...aw
(@MeowMeowHotMix)
&quot;I
35mins
&quot;Isn't
???
-
..exams.
(@isabelmariee)
.@jkmodels
5
10-Year-Old
..Get
[stapler
&quot;Hi
45Th
*yawn*
.@petraeujane
*meow*
#ivealwayswantedtosay
*sad
#FollowFriday
(@gbsinkers)
.......
(sigh)
6:29
43416
's
6
#BSNL,
50
#followfriday
*pokes
#FF
#itsucks
*YAAAWN*
�anisalovesu
31
.@orbitaldiamonds
.
...just
#Vancouver
*eekkk*
#hollyoaks
1st
*old
#twpp
8
????,
$4
...what
*uh*
#quitsmokingdiary
#Java
#twitterfails
19
*sighs*
17
(2/2)
-w

<i><b> ANALYSIS </b></i>
<p>
1. remove words starting with '@' <br>
2. remove words starting with '#'<br>
3. remove words starting with '.' / '..' etc.<br>
</p>
<p> Basically remove words that dont start with alphanumeric characters</p>

# 6. Clean the dataset

In [5]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 59281 entries, 0 to 59280
Data columns (total 2 columns):
 #   Column  Non-Null Count  Dtype 
---  ------  --------------  ----- 
 0   text    59281 non-null  object
 1   label   59281 non-null  int64 
dtypes: int64(1), object(1)
memory usage: 926.4+ KB


In [6]:
# for reference
print("ASCII for 'a': {} \nASCII for 'z': {} \nASCII for 'A': {} \nASCII for 'Z': {}". format(ord('a'), ord('z'), ord('A'), ord('Z')))
print("\n\nASCII for 0: {} \nASCII for 9: {}". format(ord('0'), ord('9')))

ASCII for 'a': 97 
ASCII for 'z': 122 
ASCII for 'A': 65 
ASCII for 'Z': 90


ASCII for 0: 48 
ASCII for 9: 57


# 6.1 Repeat the below steps till the cell below shows 0

In [73]:
row_index = list() # list of rows that have no value after removing unwanted words
not_ok_count = 0
for i in range(len(df)):
    start = ord(df['text'][i][0])
    sentence = df['text'][i]
    if (start >= 48 and start <= 57) or (start >= 97 and start <= 122) or (start >= 65 and start <= 90):
        continue
    else:
        not_ok_count += 1
        print(i, end= "\t")
        try:
            first_space = 0
            if first_space == 0:
                first_space = df['text'][i].index(" ")

            print(first_space)
            new_sentence = sentence[first_space + 1:]
            print(sentence, new_sentence, sep = "\t")
            df['text'][i] = new_sentence
        except:
            print("NEED TO REMOVE ROW ", i)
            row_index.append(i)

print(not_ok_count)
        

0


In [69]:
# list of rows to remove
print(row_index)

[]


In [70]:
df.drop(index=row_index, axis = 0, inplace = True)
df = df.reset_index()
df.drop(['index'], axis = 1, inplace = True)
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 59176 entries, 0 to 59175
Data columns (total 2 columns):
 #   Column  Non-Null Count  Dtype 
---  ------  --------------  ----- 
 0   text    59176 non-null  object
 1   label   59176 non-null  int64 
dtypes: int64(1), object(1)
memory usage: 924.8+ KB


In [ ]:
# to view as csv file for debugging
df.to_csv('debug.csv')

# 7. Save Dataset
<p> cleaning is completed</p>

In [74]:
df

text  label
0      im feeling rather rotten so im not very ambiti...      0
1              im updating my blog because i feel shitty      0
2      i never make her separate from me because i do...      0
3      i left with my bouquet of red and yellow tulip...      1
4        i was feeling a little vain when i did this one      0
...                                                  ...    ...
59171                      Succesfully following Tayla!!      1
59172                     Happy Mothers Day  All my love      2
59173  Happy Mother's Day to all the mommies out ther...      2
59174  WASSUP BEAUTIFUL!!! FOLLOW ME!!  PEEP OUT MY N...      1
59175  bullet train from tokyo    the gf and i have b...      2

[59176 rows x 2 columns]

In [75]:
df.to_csv('DL_Dataset_final.csv')